File: src/src/services/attendanceService.ts
import api from "./api"
import type { Attendance } from "../types"

const BASE_PATH = "/attendance"

export const getAttendance = async (): Promise<Attendance[]> => {
  const response = await api.get<Attendance[]>(BASE_PATH)
  return response.data
}

export const getAttendanceById = async (id: number): Promise<Attendance> => {
  const response = await api.get<Attendance>(`${BASE_PATH}/${id}`)
  return response.data
}

export const getAttendanceByStudent = async (studentId: number): Promise<Attendance[]> => {
  const response = await api.get<Attendance[]>(`${BASE_PATH}/students/${studentId}`)
  return response.data
}

export const getAttendanceByTeacher = async (teacherId: number): Promise<Attendance[]> => {
  const response = await api.get<Attendance[]>(`${BASE_PATH}/teachers/${teacherId}`)
  return response.data
}

export const getAttendanceByDate = async (date: string): Promise<Attendance[]> => {
  const response = await api.get<Attendance[]>(`${BASE_PATH}/date/${date}`)
  return response.data
}

export const getStudentAttendanceBetweenDates = async (
  studentId: number,
  startDate: string,
  endDate: string,
): Promise<Attendance[]> => {
  const response = await api.get<Attendance[]>(
    `${BASE_PATH}/students/${studentId}/between?startDate=${startDate}&endDate=${endDate}`,
  )
  return response.data
}

export const getTeacherAttendanceBetweenDates = async (
  teacherId: number,
  startDate: string,
  endDate: string,
): Promise<Attendance[]> => {
  const response = await api.get<Attendance[]>(
    `${BASE_PATH}/teachers/${teacherId}/between?startDate=${startDate}&endDate=${endDate}`,
  )
  return response.data
}

export const createAttendance = async (attendance: Attendance): Promise<Attendance> => {
  const response = await api.post<Attendance>(BASE_PATH, attendance)
  return response.data
}

export const updateAttendance = async (id: number, attendance: Attendance): Promise<Attendance> => {
  const response = await api.put<Attendance>(`${BASE_PATH}/${id}`, attendance)
  return response.data
}

export const deleteAttendance = async (id: number): Promise<void> => {
  await api.delete(`${BASE_PATH}/${id}`)
}

File: src/main/java/com/attendance/service/AttendanceService.java
package com.attendance.service;

import com.attendance.model.Attendance;
import com.attendance.model.Student;
import com.attendance.model.Teacher;
import com.attendance.repository.AttendanceRepository;
import com.attendance.repository.StudentRepository;
import com.attendance.repository.TeacherRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.LocalDate;
import java.util.List;
import java.util.Optional;

@Service
public class AttendanceService {

    private final AttendanceRepository attendanceRepository;
    private final StudentRepository studentRepository;
    private final TeacherRepository teacherRepository;

    @Autowired
    public AttendanceService(
            AttendanceRepository attendanceRepository,
            StudentRepository studentRepository,
            TeacherRepository teacherRepository) {
        this.attendanceRepository = attendanceRepository;
        this.studentRepository = studentRepository;
        this.teacherRepository = teacherRepository;
    }

    @Transactional(readOnly = true)
    public List<Attendance> getAllAttendance() {
        return attendanceRepository.findAll();
    }

    @Transactional(readOnly = true)
    public Optional<Attendance> getAttendanceById(Long id) {
        return attendanceRepository.findById(id);
    }

    @Transactional(readOnly = true)
    public List<Attendance> getAttendanceByStudent(Long studentId) {
        Optional<Student> student = studentRepository.findById(studentId);
        return student.map(attendanceRepository::findByStudent).orElse(List.of());
    }

    @Transactional(readOnly = true)
    public List<Attendance> getAttendanceByTeacher(Long teacherId) {
        Optional<Teacher> teacher = teacherRepository.findById(teacherId);
        return teacher.map(attendanceRepository::findByTeacher).orElse(List.of());
    }

    @Transactional(readOnly = true)
    public List<Attendance> getAttendanceByDate(LocalDate date) {
        return attendanceRepository.findByDate(date);
    }

    @Transactional(readOnly = true)
    public List<Attendance> getStudentAttendanceBetweenDates(Long studentId, LocalDate startDate, LocalDate endDate) {
        Optional<Student> student = studentRepository.findById(studentId);
        return student.map(s -> attendanceRepository.findByStudentAndAttendanceSession_DateBetween(s, startDate, endDate))
                .orElse(List.of());
    }

    @Transactional(readOnly = true)
    public List<Attendance> getTeacherAttendanceBetweenDates(Long teacherId, LocalDate startDate, LocalDate endDate) {
        Optional<Teacher> teacher = teacherRepository.findById(teacherId);
        return teacher.map(t -> attendanceRepository.findByTeacherAndAttendanceSession_DateBetween(t, startDate, endDate))
                .orElse(List.of());
    }

    @Transactional
    public Attendance saveAttendance(Attendance attendance) {
        // Si es asistencia de estudiante, verificar que el estudiante exista
        if (attendance.getType() == Attendance.AttendanceType.STUDENT && attendance.getStudent() != null) {
            Optional<Student> studentOpt = studentRepository.findById(attendance.getStudent().getId());
            if (studentOpt.isPresent()) {
                attendance.setStudent(studentOpt.get());
            }
        }

        // Si es asistencia de profesor, verificar que el profesor exista
        if (attendance.getType() == Attendance.AttendanceType.TEACHER && attendance.getTeacher() != null) {
            Optional<Teacher> teacherOpt = teacherRepository.findById(attendance.getTeacher().getId());
            if (teacherOpt.isPresent()) {
                attendance.setTeacher(teacherOpt.get());
            }
        }

        return attendanceRepository.save(attendance);
    }

    @Transactional
    public void deleteAttendance(Long id) {
        attendanceRepository.deleteById(id);
    }
}

